### PR TITLE
Add support for inline image uploads to telegram notifier plugin

### PR DESCRIPTION
**Pull request by rburchell**
As requested in #8246, and initially started in #8373 by @wf1nder, add support for inline image uploads to the telegram notifier plugin.

This implementation is different:
* rather than directly using net/http, it still uses SendWebhookSync in an attempt to reuse more of core functionality
* as a result, rather than using the existing simplejson payload, it uses multipart for both text and image uploads
* there are (effectively) no changes to message content included in this pull request, with the exception that image messages do not have HTML tags as they do not appear to be supported. Any changes to message content should be done as a separate pull request.

There is still some slight duplication in how the payloads are created, but in my opinion, it likely isn't worth the effort to share any of this code: image and text messages have different API with different parameters, and it isn't a large amount of code, after all.

No unit tests have been added at this time, as it is not clear to me precisely what should be tested, and how, especially as this would require connection to a live external service.

### CODE REVIEW

**Comment by rburchell** [comment]
@torkelo I'm still not really sure what improvements precisely you wanted to see, but I've made an attempt to mind read. :)

Unfortunately, I have no idea what kind of unit testing you'd want, and looking at the other notifier tests, I wasn't able to get any real inspiration. They all look very minimal, and just test settings, which is what I'd expect given that they by and large rely on external services, including e.g. the Slack notifier which has a similar upload image feature.

**Comment by rburchell** [comment]
Screenshots. Old public URL notification first, new inline image notification second.

image

File: pkg/services/alerting/notifiers/telegram.go
Comment (bergquist) [inline pkg/services/alerting/notifiers/telegram.go]: This should be an else statement for `evalContext.ImagePublicUrl == "" && this.UploadImage == true`

File: pkg/services/alerting/notifiers/telegram.go
Comment (bergquist) [inline pkg/services/alerting/notifiers/telegram.go]: Too much code duplication in `buildImageMessage` and `buildTextMessage` 

Rather then having two functions doing almost the same thing. Have one function and add image fields if send image is set to true.

**Review by bergquist — CHANGES_REQUESTED** [review]

**Comment by bergquist** [comment]
Thank you for taking the time to contribute to Grafana. We would like to see a few changes before merging this.

File: pkg/services/alerting/notifiers/telegram.go
Comment (rburchell) [reply pkg/services/alerting/notifiers/telegram.go]: The problem is that they are "almost the same thing", not "the same thing".

Differences:
* buildImageMessage needs to not embed HTML tags into the message (and it's called "caption" in the form data, not "message")
* one sends to sendPhoto, the other to sendMessage URI
* one embeds a file, the other does not

I think that merging them together will look subjectively uglier, but I'll do it as it's not a big deal, and you want that. I'll push an update this evening.

**Comment by rburchell** [comment]
Merged into one function as requested. Let me know if there's anything more.

**Comment by rburchell** [comment]
Ping? I'm happy to revise this further, but I need direction to do so.

**Comment by rburchell** [comment]
@bergquist Ping. I'm sorry to keep nudging, but I'd like to get this done before everything (myself included) shuts down over Christmas and this gets forgotten.

**Comment by rburchell** [comment]
I'm happy to fix the merge conflicts, but I'd like to know first if anything else is required.

**Comment by rburchell** [comment]
Ping!

(and happy new year 🎆 )

**Comment by fanvinga** [comment]
Have this feature been merged into master branch and availd in 4.6.3 release ?

**Comment by rburchell** [comment]
@fanvinga Not yet, as far as I know. I'm still waiting for feedback.

**Comment by fanvinga** [comment]
Can't wait to see any progress  :)
Telegram Bot is really convenient :)

**Comment by wizardjedi** [comment]
Hi, folks.

This feature is really important. Maybe it's time to merge it and release?!

**Review by bergquist — APPROVED** [review]

**Comment by bergquist** [comment]
Thank you for the patients. Setting up a telegram bot for testing this is quite a big task. @wizardjedi @fanvinga any chance anyone of you want to build this and test it?

**Comment by wizardjedi** [comment]
@bergquist Hi, man. I'll try to make a test. But i'm noob in building go apps, but I'm going to try in my production environment.

**Comment by fanvinga** [comment]
@bergquist Hi,man. The way I using grafana is through Docker . I will try to build a new docker image for that to test it

**Comment by fanvinga** [comment]
@bergquist Hey,man. I just have cloned rburchell's grafana fork and used official container build script to build a new image for using. For privacy reasons , I can't post the sample image , however it actually works well.My bot successfully sent me a alerting with inline image . :)  Such a good job rburchell've done :)

**Comment by bergquist** [comment]
@fanvinga thanks

@rburchell Thanks for your patience. Please rebase this branch and I will merge it.

**Comment by rburchell** [comment]
@bergquist Done, and no worries, I understand everyone has other things to do. Thanks for taking a look!

@fanvinga Thanks for testing!

**Comment by fanvinga** [comment]
@rburchell @bergquist You're welcome, making grafana better is our common wish. :)

**Comment by bergquist** [comment]
Thank you for contributing to Grafana!

This will be available in 5.0. the master build will be ready in ~30min at grafana.com :)

**Comment by wizardjedi** [comment]
@fanvinga @rburchell @bergquist I've tested nightly build too. It works fine.

**Comment by fanvinga** [comment]
@rburchell Hey , man . It seems that there are some conflicts in newest commits . The feature of upload inline image through Telegram Bot doesn't work in commit 39dd479. I guess that the local image uploader (new feature) cause it . Can you check for it ?

**Comment by fanvinga** [comment]
7f1b58a4-0831-4e05-bb64-c0ca2aa3a47e
Just like that now :(

**Comment by rburchell** [comment]
@fanvinga I'm not going to be able to spend much time on this myself until maybe April, due to other projects, but if you point me at what caused the break I can try figure out how to fix it if it's not too bad.

Can you bisect the breakage?

**Comment by fanvinga** [comment]
@rburchell I am busy with my final examination these days :( 
But maybe I can try to find out the problem next weekend ? 
Because at that time I may have a two-week holiday :)

**Comment by bergquist** [comment]
remove local fileuploader as the default option. I think that should solve this problem.

```
[external_image_storage]
# You can choose between (s3, webdav, gcs, azure_blob, local)
provider =
```

**Comment by fanvinga** [comment]
@bergquist I noticed that you added a commit that remove the default image uploader. So I pull the latest master image to use. However it seems doesn't work :( The alerting notifications now become like that .
d04e5016-8130-490b-a125-afe026cf303b

**Comment by bergquist** [comment]
@fanvinga you might have to replace `provider = local` with `provider = ` in grafana.ini yourself

**Comment by fanvinga** [comment]
1cd8e336-84b2-46a1-b97e-bbe741ce7c1c
@bergquist Thanks for your advice. I checked my grafana.ini and saw that this actually blank. And I edit the config file like the picture I post then restart grafana . The situation still the same as my last try. :(

**Comment by bergquist** [comment]
Lets move the conversation back to https://github.com/grafana/grafana/issues/8246

**Comment by xlson** [comment]
Continued discussions regarding inline images in #10631

**Comment by IamAliBaba** [comment]
Hi all,

there is issue during local attaching process.

[https://cloud.bipmedia.com/index.php/s/t8nEgL3AnLi2w3q](url)

**Comment by IamAliBaba** [comment]
Hello @wizardjedi 

What version of grafana are u using ?

Can you send your grafana.ini configs regarding to alerts ?

Thanks,
Ali

**Pull request merged**
